### PR TITLE
Initial version of the audio service

### PR DESCRIPTION
**Pull request by sturufous**
This is the audio ingestion service cloned from Jeremy's RSS service. While it is functional and meets all of the acceptance criteria, I may have missed some of the subtleties of how the service infrastructure operates. For instance, I'm unclear on how futures work with Kafka and may not have handled those properly.

This implementation reads all configuration information from audio.env and enters a loop that repeats every minute. This delay is short to allow the solution to be demonstrated by extracting multiple clips over a short period of time. The service merges the Capture and Clip logic of Jorel into one event. If the capture file is missing, or older than one minute, the event will be interpreted as a begin-capture event. (An old capture file indicates that the audio stream has dropped.) 

As a clip can't be extracted from the capture file until audio representing the length of the clip has been recorded, the capture and clip events are mutually exclusive (the service doesn't attempt to clip from a file into which recording has just begun). If the capture file doesn't contain enough audio to represent the clip, ffmpeg creates a (mostly) empty clip file. The clip length is set to 50 seconds so that there is guaranteed to be enough audio to extract between the begin-capture and clip event. 

Captured audio is written to a file with the name of the source (in this case CBCKAM.mpg - CBC Kamloops) in the capture directory configured. Clips are written to a sub-directory of the configured clips directory. The directory names are of the format:

`tno/clips/20220104/CBCKAM`

These directories are created automatically if they do not already exist. Eventually, these directories will be located on storage that is shared by the audio producer and consumer. The final solution could use an "execute at <time>" event for each clip, which should reproduce the schedule functionality of the Jorel clip event. One issue to consider is that the audio clips are not the final product that will be processed and published. In Jorel these files are called clips whereas the edited and published audio files are called snippets.

### CODE REVIEW

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      cmd = cmd.replace("[capture-path]", captureFilePath);
+
+      String[] cmda = {
+        "/bin/sh",
+            "-c",
+        cmd
+      };
+
+      try {
+        Process p = new ProcessBuilder(cmda).start();
```
Comment (Fosol) [inline services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java:123]: Does running this process block the service thread?  Will a failure in the process after it has started bubble up into this try+catch"?

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+   * @param cmd The Linux command to execute 
+   * @param captureFilePath The full path of the output directory
+   */
+  private void executeCaptureCmd(String cmd, String captureFilePath) {
+
+    if (!cmd.equals("")) {
+      cmd = cmd.replace("[capture-url]", audioConfig.getAudioUrl());
+      cmd = cmd.replace("[capture-path]", captureFilePath);
+
+      String[] cmda = {
```
Comment (Fosol) [inline services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java:116]: Does it make sense to use a java library to interface with ffmpeg?  For example - https://github.com/bytedeco/javacpp-presets/tree/master/ffmpeg

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/kafka/KafkaAudioProducer.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  public void send(ProducerSendEvent event) throws InterruptedException, ExecutionException {
+    var clip = event.getData();
+
+    var config = event.getConfig();
+    var source = config.getId();
+    var topic = config.getTopic();
+
+    try {
+      logger.info(String.format("New audio content extracted: '%s', topic: %s", clip, topic));
+      var contentReference = new ContentReference(source, clip, topic);
```
Comment (Fosol) [inline services/audio/src/main/java/ca/bc/gov/tno/services/audio/kafka/KafkaAudioProducer.java:108]: This ContentReference needs to somehow point to the Clip that was created.  It probably should be just the clip name.  That way the Transcription Service can grab the clip once we build it.

**Review by Fosol — COMMENTED** [review]
Update the `./tools/scripts/gen-env-files.sh` to create a valid `audio.env` file for this service.
Update the `./docker-compose.yml` file to include the audio service.

File: services/audio/pom.xml [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-databind</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>com.fasterxml.jackson.dataformat</groupId>
+      <artifactId>jackson-dataformat-yaml</artifactId>
+    </dependency>
+
+    <dependency>
```
[the remark this answers is not in the document]
Comment (sturufous) [reply services/audio/pom.xml:119]: I can remove this. However, I noticed that in the pom.xml for the syndication service all the Jackson dependencies are duplicated. As I based my pom on that one I also have duplications. Is there a reason for this?

[the remark this answers is not in the document]
Comment (Fosol) [reply services/audio/pom.xml:119]: Their version is defined within the <dependencyManagement>.  I'm not an expert in java, I simply followed examples I found online.  If these can be cleaned up and still work correctly, feel free to do so.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/BaseScheduleService.class [hunk 5]
```diff
@@ -0,0 +1,208 @@
+package ca.bc.gov.tno.services.data;
```
[the remark this answers is not in the document]
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/BaseScheduleService.class:1]: No I've never seen it done. .class is the compiled byte-code for the .java file. It seems that when I run the service in the vscode debugger it pulls the BaseScheduleService.class file out of our library and puts it in the handlers directory. I had intended to add .class files to the .gitignore file but it slipped my mind. I'll do that.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/BaseScheduleService.class [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+/**
+ * BaseScheduleService class, provides a process that manages the scheduling of
+ * ingestion. This will ensure the process runs according to the configured
+ * schedule.
+ * 
+ * @param <C>  The data source config type.
+ * @param <CA> The data source config collection type.
+ */
+@Async
+public abstract class BaseScheduleService<C extends DataSourceConfig, CA extends DataSourceCollectionConfig<C>>
```
[the remark this answers is not in the document]
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/BaseScheduleService.class:31]: See above.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/BaseScheduleService.class (hunk 5 above)
[the remark this answers is not in the document]
Comment (Fosol) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/BaseScheduleService.class:1]: Why is it dumping `.class` files in these folders?  Seems like strange behaviour to me.  I'd look into it.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/config/AudioConfig.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+ * Configuration settings for the default Audio Feed. These values will be
+ * used if a connection to the database cannot be made.
+ */
+@Configuration
+//@ConfigurationProperties("data")
+public class AudioConfig extends DataSourceConfig {
+  /**
+   * URL to the audio data source feed
+   */
+  private String url;
```
[the remark this answers is not in the document]
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/config/AudioConfig.java:21]: It's not really, but I preferred the name audioUrl. I've deleted the url variable and it's setter/getter.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    if (!captureFile.exists()) {
+        executeCaptureCmd(cmd, captureFilePath);
+        audioConfig.setStreamStartTime(System.currentTimeMillis());
+        startedStream = true;
+    } else {
+      // Check to see if the audio stream has dropped
+      // TODO: handle rollover from one day to next
+      long modified = captureFile.lastModified();
+      long current = System.currentTimeMillis();
+      if (current - modified > 60000) {
```
[the remark this answers is not in the document]
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java:94]: I added the setting streamTimeout

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/BaseScheduleService.class [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+/**
+ * BaseScheduleService class, provides a process that manages the scheduling of
+ * ingestion. This will ensure the process runs according to the configured
+ * schedule.
+ * 
+ * @param <C>  The data source config type.
+ * @param <CA> The data source config collection type.
+ */
+@Async
+public abstract class BaseScheduleService<C extends DataSourceConfig, CA extends DataSourceCollectionConfig<C>>
```
[the remark this answers is not in the document]
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/BaseScheduleService.class:31]: Already discussed.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java (hunk 1 above)
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java:123]: The process variable p is returned without blocking. It can be used to observe the process and see why it has terminated. The existing system does nothing with these variables. A failure in the process can return an exit code but is unlikely to throw an exception.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java (hunk 2 above)
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java:116]: I can look into this after the demo tomorrow.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+
+    audioConfig = event.getConfig();
+    String mediaSource = audioConfig.getCaptureDir() + "/" + audioConfig.getId() + ".mpg";
+    String cmd = audioConfig.getClipCmd();
+    String destination = clipPath();
+
+    try {
+      caller = event.getSource();
+      var url = audioConfig.getAudioUrl();
+      var readyEvent = new ProducerSendEvent(caller, audioConfig, destination);
```
[the remark this answers is not in the document]
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java:58]: I applied this change.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+    this.eventPublisher = eventPublisher;
+  }
+
+  /**
+   * Start an audio stream, if not already in progress, and extract an audio clip. Publish an audio event so that the Kafka 
+   * Producer can push the content to a topic.
+   */
+  @Override
+  @EventListener
+  public void onApplicationEvent(TransactionBeginEvent<AudioConfig> event) {
```
[the remark this answers is not in the document]
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java:48]: This was my prototypical implementation of your suggestion in our Teams chat from December 27:

`As discussed before, ideally your Audio Streaming Service will first capture the stream and create a file. Then a separate event within the same service will handle the scheduling of the Clip event.`

I just wanted to grab the media and proof the clipping process. The existing system has a separate event type and handler for capture and clip events. We can discuss how that might play out in our new design after the sprint review.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+    }
+  }
+
+  /**
+   * Obtain the location of the streaming file for this event and check that it is being written to.
+   * If the file doesn't exist, or the audio stream has dropped, start streaming from the media source.
+   */
+  private boolean startAudioStream() {
+
+    String captureFilePath = audioConfig.getCaptureDir() + "/" + audioConfig.getId() + ".mpg";
```
[the remark this answers is not in the document]
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java:78]: Applied the change.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+    File captureFile = new File(captureFilePath);
```
[the remark this answers is not in the document]
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java:79]: java.io.File is just an abstract representation of a file path rather than an open file that you can read. The file doesn't require closing and will be garbage collected after startAudioStream() is popped off the stack.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+        executeCaptureCmd(cmd, captureFilePath);
+        audioConfig.setStreamStartTime(System.currentTimeMillis());
+        startedStream = true;
+    } else {
+      // Check to see if the audio stream has dropped
+      // TODO: handle rollover from one day to next
+      long modified = captureFile.lastModified();
+      long current = System.currentTimeMillis();
+      if (current - modified > 60000) {
+        executeCaptureCmd(cmd, captureFilePath);
```
[the remark this answers is not in the document]
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/handlers/FetchDataService.java:95]: Addressed this.

File: services/audio/src/main/java/ca/bc/gov/tno/services/audio/kafka/KafkaAudioProducer.java (hunk 3 above)
Comment (sturufous) [reply services/audio/src/main/java/ca/bc/gov/tno/services/audio/kafka/KafkaAudioProducer.java:108]: At the moment I have no uid to put in the uid column of ContentReference, so I'm storing the clip-path as a unique value in the uid field. I can shorten this to the clip name or discuss what you think would be a better solution.

File: services/audio/.gitignore [hunk 15]
```diff
@@ -0,0 +1,38 @@
+HELP.md
+target/
+!.mvn/wrapper/maven-wrapper.jar
+!**/src/main/**/target/
+!**/src/test/**/target/
+/logs
+*.class
```
Comment (Fosol) [inline services/audio/.gitignore:7]: Not certain this makes a lot of sense, but it'll do for now.  I'd like to know what generates these files.

**Review by Fosol — APPROVED** [review]
